### ISSUE TITLE
bigquery: retry on error code:500

### ISSUE THREAD

**Issue by careylam** [googleapis/google-cloud-go#5248]
**Client**

BigQuery

**Environment**

MacOS

**Go Environment**

go version go1.17.2 darwin/amd64

$ go env

**Code**

e.g.
```go
func retryableError(err error) bool {
	if err == nil {
		return false
	}
	if err == io.ErrUnexpectedEOF {
		return true
	}
	// Special case due to http2: https://github.com/googleapis/google-cloud-go/issues/1793
	// Due to Go's default being higher for streams-per-connection than is accepted by the
	// BQ backend, it's possible to get streams refused immediately after a connection is
	// started but before we receive SETTINGS frame from the backend.  This generally only
	// happens when we try to enqueue > 100 requests onto a newly initiated connection.
	if err.Error() == "http2: stream closed" {
		return true
	}

	switch e := err.(type) {
	case *googleapi.Error:
		// We received a structured error from backend.
		var reason string
		if len(e.Errors) > 0 {
			reason = e.Errors[0].Reason
		}
		if e.Code == http.StatusServiceUnavailable || e.Code == http.StatusBadGateway || reason == "backendError" || reason == "rateLimitExceeded" {
			return true
		}
	case *url.Error:
		retryable := []string{"connection refused", "connection reset"}
		for _, s := range retryable {
			if strings.Contains(e.Error(), s) {
				return true
			}
		}
	case interface{ Temporary() bool }:
		if e.Temporary() {
			return true
		}
	}
	// Unwrap is only supported in go1.13.x+
	if e, ok := err.(interface{ Unwrap() error }); ok {
		return retryableError(e.Unwrap())
	}
	return false
}

```

**Expected behavior**
When e.Code == http.StatusInternalServerError, according to the error message from BigQuery,  it should retry. 

**Actual behavior**
The error is not retried and return directly. Here's a sample message from our production env. 
```
"error":"googleapi: Error 500: An internal error occurred and the request could not be completed. This is usually caused by a transient issue. Retrying the job with back-off as described in the BigQuery SLA should solve the problem: https://cloud.google.com/bigquery/sla. If the error continues to occur please contact support at https://cloud.google.com/support., internalError"
```

**Comment by shollyman** [comment]
The `retryableError` predicate is used for retrying API method interactions (e.g. starting a query, getting table metadata, etc).

In this case, it seems like you started a job, and the methods for interacting with the job did not have issue (inserting it, polling it for status, etc).  However, the job itself bears an error state from failing execution.  Once a job completes it is final, and can't be retried.  A new job can be created using the same configuration and run, but this is not currently handled by the library.

We've avoided doing job recreations automatically as there's some complexities around safely retrying, and thus this is currently a user level retry concern.  Can you provide an example about how you're invoking jobs/queries, and possibly a little more detail about the nature of the job/query itself (e.g. a script, or a SELECT query, or load job, etc)?

**Comment by careylam** [comment]
Our call is simple, just call the following func from insert.go
```go
func (u *Inserter) Put(ctx context.Context, src interface{}) (err error)
```

based on message return from BigQuery and the page: https://cloud.google.com/bigquery/docs/error-messages, BigQuery should retry it automatically. Right? 

Here's my fix:
```go
		if e.Code == http.StatusServiceUnavailable || e.Code == http.StatusBadGateway || e.Code == http.StatusInternalServerError ||
			reason == "internalError" || reason == "backendError" || reason == "rateLimitExceeded" {
			return true
		}
```
What do you think?

**Comment by shollyman** [comment]
The `Inserter` abstraction is used for streaming data.  If you're getting an internalError from the backend I'd expect it to retry with the existing predicate as the 'internalError' reason should be extracted from the structured json error response.

The error in your original report is a job-related error, which doesn't involve the streaming API.  Are you getting that from the Inserter, or is another error surfacing?

**Comment by careylam** [comment]
Sorry, I forgot to get this clear. The error is logged from our application depends on cloud.google.com/go/bigquery v1.8. It happens everyday. In that version, we are streaming the data by using the <mark>bigquery.Uploader</mark>. 
```go
	uploader := config.Client.
		DatasetInProject(config.ProjectName, config.DatasetName).
		Table(tableNameWithPartition).
		Uploader()
```
I am in the process to upgrade it to the latest v1.25.0. So, I am validating if HTTP 500 is retried in v1.25.0. The unit test of <mark>retryableError</mark> has a test case to expect HTTP 500 not to retry. That's why I am creating this ticket.
```go
		{
			// not retried per https://google.aip.dev/194
			"internal error",
			&googleapi.Error{
				Code: http.StatusInternalServerError,
			},
			false,
		},
```

I suspect that HTTP 500 will not retry after I upgrade to v1.25.0. Thoughts?

BTW, I tried to push my fix the other day. It seems I am allowed. How can I become a contributor? Thank you

**Comment by shollyman** [comment]
Yes, per the AIP guidance blindly retrying on http 500 is not recommended.  However, a 500 response that includes a structured error may be retried, that's what the `retryableError` function is responsible for evaluating.  The error processing in 1.8.0 is less aware of retryable conditions than what's available in the latest version, so it should strictly be an improvement.

**Comment by shollyman** [comment]
For posterity, here's the function from 1.8.0:
```
func retryableError(err error) bool {
	e, ok := err.(*googleapi.Error)
	if !ok {
		return false
	}
	var reason string
	if len(e.Errors) > 0 {
		reason = e.Errors[0].Reason
	}
	return e.Code == http.StatusServiceUnavailable || e.Code == http.StatusBadGateway || reason == "backendError" || reason == "rateLimitExceeded"
}
```

**Comment by careylam** [comment]
Are you going to fix it to retry on a recoverable 500?

**Comment by careylam** [comment]
Would it make sense to allow an error handler callback? so the client can enhance the error handling on their own use case. In my example, there are a few more that we want to handle it, e.g. 401, 403, HTTP client no usable, context deadline limit...

**Comment by dominicbarnes** [comment]
When using the `Inserter`, I'm experiencing 500 errors infrequently (but enough to cause some CI builds to fail):

> googleapi: Error 500: An internal error occurred and the request could not be completed. This is usually caused by a transient issue. Retrying the job with back-off as described in the BigQuery SLA should solve the problem: https://cloud.google.com/bigquery/sla. If the error continues to occur please contact support at https://cloud.google.com/support., internalError

As was pointed out previously, even the error message suggests it should be retried, so it seems within the scope of this client library to do this transparently. I'm using a more recent version of this SDK (v1.34.1) and it appears that "internalError" is still not a [retryable reason](https://github.com/googleapis/google-cloud-go/blob/72ccadaeef1ace298366f16c205e5242695159f9/bigquery/bigquery.go#L199), nor is status code 500 [considered retryable](https://github.com/googleapis/google-cloud-go/blob/72ccadaeef1ace298366f16c205e5242695159f9/bigquery/bigquery.go#L235). (just 502 & 503)

I was contemplating just wrapping this with my own retry logic, but was curious if a PR would be accepted for changing either of these? It sounds like 500 is probably not safe to blindly retry, but maybe a case can be made for "internalError" as one of the possible reasons? (or is "internalError" just as ambiguous as 500?)

**Comment by maruel** [comment]
This issue affects both Fuchsia and Chrome infrastructures.

@shollyman , https://cloud.google.com/bigquery/docs/error-messages#errortable is very clear: `If you receive a 5xx response code, then retry the request later.` It's not talking about specific 500 codes, it says all of them. See cl/339885876 internally.

I don't understand the resistance here. You take AIP directive over your own product's documentation. Can you either update the official documentation or update the retry algorithm. The fact that the retry mechanism is not configurable (including the timeout) is a concern for us too. We will end up retrying anyway.

**Comment by codyoss** [comment]
@shollyman is there something actionable to do with this issue?

**Comment by vegather** [comment]
Looks like there was an attempt to fix this in `1.44.0`, but I'm still seeing the same issue in `1.57.1` when using the `Put` method on an `Inserter`.
